File: examples/android/src/main/java/com/dropbox/core/examples/android/FilesActivity.java
package com.dropbox.core.examples.android;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.webkit.MimeTypeMap;
import android.widget.Toast;

import com.dropbox.core.v2.files.FileMetadata;
import com.dropbox.core.v2.files.FolderMetadata;
import com.dropbox.core.v2.files.ListFolderResult;

import java.io.File;
import java.text.DateFormat;
import java.util.List;


/**
 * Activity that displays the content of a path in dropbox and lets users navigate folders,
 * and upload/download files
 */
public class FilesActivity extends DropboxActivity {
    private static final String TAG = FilesActivity.class.getName();

    public final static String EXTRA_PATH = "FilesActivity_Path";
    private static final int PICKFILE_REQUEST_CODE = 1;

    private String mPath;
    private FilesAdapter mFilesAdapter;
    private FileMetadata mSelectedFile;

    public static Intent getIntent(Context context, String path) {
        Intent filesIntent = new Intent(context, FilesActivity.class);
        filesIntent.putExtra(FilesActivity.EXTRA_PATH, path);
        return filesIntent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        String path = getIntent().getStringExtra(EXTRA_PATH);
        mPath = path == null ? "" : path;

        setContentView(R.layout.activity_files);

        Toolbar toolbar = (Toolbar) findViewById(R.id.app_bar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = (FloatingActionButton)findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                performWithPermissions(FileAction.UPLOAD);
            }
        });
        //init picaso client
        PicassoClient.init(this,DropboxClientFactory.getClient());
        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.files_list);
        mFilesAdapter = new FilesAdapter(PicassoClient.getPicasso(), new FilesAdapter.Callback() {
            @Override
            public void onFolderClicked(FolderMetadata folder) {
                startActivity(FilesActivity.getIntent(FilesActivity.this, folder.getPathLower()));
            }

            @Override
            public void onFileClicked(final FileMetadata file) {
                mSelectedFile = file;
                performWithPermissions(FileAction.DOWNLOAD);
            }
        });
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(mFilesAdapter);

        mSelectedFile = null;
    }

    private void launchFilePicker() {
        // Launch intent to pick file for upload
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("*/*");
        startActivityForResult(intent, PICKFILE_REQUEST_CODE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, final Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == PICKFILE_REQUEST_CODE) {
            if (resultCode == RESULT_OK) {

                // This is the result of a call to launchFilePicker
                uploadFile(data.getData().toString());
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int actionCode, @NonNull String [] permissions, @NonNull int [] grantResults) {
        FileAction action = FileAction.fromCode(actionCode);

        boolean granted = true;
        for (int i = 0; i < grantResults.length; ++i) {
            if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                Log.w(TAG, "User denied " + permissions[i] +
                        " permission to perform file action: " + action);
                granted = false;
                break;
            }
        }

        if (granted) {
            performAction(action);
        } else {
            switch (action) {
                case UPLOAD:
                    Toast.makeText(this,
                            "Can't upload file: read access denied. " +
                                    "Please grant storage permissions to use this functionality.",
                            Toast.LENGTH_LONG)
                        .show();
                    break;
                case DOWNLOAD:
                    Toast.makeText(this,
                            "Can't download file: write access denied. " +
                                    "Please grant storage permissions to use this functionality.",
                            Toast.LENGTH_LONG)
                        .show();
                    break;
            }
        }
    }

    private void performAction(FileAction action) {
        switch(action) {
            case UPLOAD:
                launchFilePicker();
                break;
            case DOWNLOAD:
                if (mSelectedFile != null) {
                    downloadFile(mSelectedFile);
                } else {
                    Log.e(TAG, "No file selected to download.");
                }
                break;
            default:
                Log.e(TAG, "Can't perform unhandled file action: " + action);
        }
    }

    @Override
    protected void loadData() {

        final ProgressDialog dialog = new ProgressDialog(this);
        dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        dialog.setCancelable(false);
        dialog.setMessage("Loading");
        dialog.show();

        new ListFolderTask(DropboxClientFactory.getClient(), new ListFolderTask.Callback() {
            @Override
            public void onDataLoaded(ListFolderResult result) {
                dialog.dismiss();

                mFilesAdapter.setFiles(result.getEntries());
            }

            @Override
            public void onError(Exception e) {
                dialog.dismiss();

                Log.e(TAG, "Failed to list folder.", e);
                Toast.makeText(FilesActivity.this,
                        "An error has occurred",
                        Toast.LENGTH_SHORT)
                        .show();
            }
        }).execute(mPath);
    }

    private void downloadFile(FileMetadata file) {
        final ProgressDialog dialog = new ProgressDialog(this);
        dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        dialog.setCancelable(false);
        dialog.setMessage("Downloading");
        dialog.show();

        new DownloadFileTask(FilesActivity.this, DropboxClientFactory.getClient(), new DownloadFileTask.Callback() {
            @Override
            public void onDownloadComplete(File result) {
                dialog.dismiss();

                if (result != null) {
                    viewFileInExternalApp(result);
                }
            }

            @Override
            public void onError(Exception e) {
                dialog.dismiss();

                Log.e(TAG, "Failed to download file.", e);
                Toast.makeText(FilesActivity.this,
                        "An error has occurred",
                        Toast.LENGTH_SHORT)
                        .show();
            }
        }).execute(file);

    }

    private void viewFileInExternalApp(File result) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        MimeTypeMap mime = MimeTypeMap.getSingleton();
        String ext = result.getName().substring(result.getName().indexOf(".") + 1);
        String type = mime.getMimeTypeFromExtension(ext);

        intent.setDataAndType(Uri.fromFile(result), type);

        // Check for a handler first to avoid a crash
        PackageManager manager = getPackageManager();
        List<ResolveInfo> resolveInfo = manager.queryIntentActivities(intent, 0);
        if (resolveInfo.size() > 0) {
            startActivity(intent);
        }
    }

    private void uploadFile(String fileUri) {
        final ProgressDialog dialog = new ProgressDialog(this);
        dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        dialog.setCancelable(false);
        dialog.setMessage("Uploading");
        dialog.show();

        new UploadFileTask(this, DropboxClientFactory.getClient(), new UploadFileTask.Callback() {
            @Override
            public void onUploadComplete(FileMetadata result) {
                dialog.dismiss();

                String message = result.getName() + " size " + result.getSize() + " modified " +
                        DateFormat.getDateTimeInstance().format(result.getClientModified());
                Toast.makeText(FilesActivity.this, message, Toast.LENGTH_SHORT)
                        .show();

                // Reload the folder
                loadData();
            }

            @Override
            public void onError(Exception e) {
                dialog.dismiss();

                Log.e(TAG, "Failed to upload file.", e);
                Toast.makeText(FilesActivity.this,
                        "An error has occurred",
                        Toast.LENGTH_SHORT)
                        .show();
            }
        }).execute(fileUri, mPath);
    }

    private void performWithPermissions(final FileAction action) {
        if (hasPermissionsForAction(action)) {
            performAction(action);
            return;
        }

        if (shouldDisplayRationaleForAction(action)) {
            new AlertDialog.Builder(this)
                    .setMessage("This app requires storage access to download and upload files.")
                    .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            requestPermissionsForAction(action);
                        }
                    })
                    .setNegativeButton("Cancel", null)
                    .create()
                    .show();
        } else {
            requestPermissionsForAction(action);
        }
    }

    private boolean hasPermissionsForAction(FileAction action) {
        for (String permission : action.getPermissions()) {
            int result = ContextCompat.checkSelfPermission(this, permission);
            if (result == PackageManager.PERMISSION_DENIED) {
                return false;
            }
        }
        return true;
    }

    private boolean shouldDisplayRationaleForAction(FileAction action) {
        for (String permission : action.getPermissions()) {
            if (!ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
                return true;
            }
        }
        return false;
    }

    private void requestPermissionsForAction(FileAction action) {
        ActivityCompat.requestPermissions(
                this,
                action.getPermissions(),
                action.getCode()
        );
    }

    private enum FileAction {
        DOWNLOAD(Manifest.permission.WRITE_EXTERNAL_STORAGE),
        UPLOAD(Manifest.permission.READ_EXTERNAL_STORAGE);

        private static final FileAction [] values = values();

        private final String [] permissions;

        FileAction(String ... permissions) {
            this.permissions = permissions;
        }

        public int getCode() {
            return ordinal();
        }

        public String [] getPermissions() {
            return permissions;
        }

        public static FileAction fromCode(int code) {
            if (code < 0 || code >= values.length) {
                throw new IllegalArgumentException("Invalid FileAction code: " + code);
            }
            return values[code];
        }
    }
}

File: examples/android/src/main/java/com/dropbox/core/examples/android/FilesActivity.kt
package com.dropbox.core.examples.android

import android.Manifest
import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.View
import android.webkit.MimeTypeMap
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.dropbox.core.examples.android.internal.api.DownloadFileTaskResult
import com.dropbox.core.examples.android.internal.api.FileMetadataApiResult
import com.dropbox.core.examples.android.internal.api.ListFolderApiResult
import com.dropbox.core.examples.android.internal.ui.FilesAdapter
import com.dropbox.core.examples.android.internal.util.UriHelpers
import com.dropbox.core.v2.files.FileMetadata
import com.dropbox.core.v2.files.FolderMetadata
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.io.File
import java.text.DateFormat
import kotlinx.coroutines.launch


/**
 * Activity that displays the content of a path in dropbox and lets users navigate folders,
 * and upload/download files
 */
class FilesActivity : BaseSampleActivity() {
    private var mPath: String? = null
    private var mFilesAdapter: FilesAdapter? = null
    private var mSelectedFile: FileMetadata? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPath = intent.getStringExtra(EXTRA_PATH) ?: ""
        setContentView(R.layout.activity_files)

        val toolbar = findViewById<View>(R.id.app_bar) as Toolbar
        setSupportActionBar(toolbar)
        val fab = findViewById<View>(R.id.fab) as FloatingActionButton
        fab.setOnClickListener { performWithPermissions(FileAction.UPLOAD) }
        val recyclerView = findViewById<View>(R.id.files_list) as RecyclerView
        mFilesAdapter = FilesAdapter(
            dbxClientV2 = dropboxApiWrapper.dropboxClient,
            mCallback = object : FilesAdapter.Callback {
                override fun onFolderClicked(folder: FolderMetadata?) {
                    requireNotNull(folder)
                    startActivity(getIntent(this@FilesActivity, folder.pathLower))
                }

                override fun onFileClicked(file: FileMetadata?) {
                    mSelectedFile = file
                    performWithPermissions(FileAction.DOWNLOAD)
                }
            },
            scope = lifecycleScope,
        )
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = mFilesAdapter
        mSelectedFile = null
    }

    private fun launchFilePicker() {
        // Launch intent to pick file for upload
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        intent.type = "*/*"
        startActivityForResult(intent, PICKFILE_REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == PICKFILE_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {

                // This is the result of a call to launchFilePicker
                uploadFile(data!!.data.toString())
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        val action = FileAction.fromCode(requestCode)
        var granted = true
        for (i in grantResults.indices) {
            if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                Log.w(
                    TAG, "User denied " + permissions[i] +
                            " permission to perform file action: " + action
                )
                granted = false
                break
            }
        }
        if (granted) {
            performAction(action)
        } else {
            when (action) {
                FileAction.UPLOAD -> Toast.makeText(
                    this,
                    "Can't upload file: read access denied. " +
                            "Please grant storage permissions to use this functionality.",
                    Toast.LENGTH_LONG
                )
                    .show()
                FileAction.DOWNLOAD -> Toast.makeText(
                    this,
                    "Can't download file: write access denied. " +
                            "Please grant storage permissions to use this functionality.",
                    Toast.LENGTH_LONG
                )
                    .show()
            }
        }

        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun performAction(action: FileAction) {
        when (action) {
            FileAction.UPLOAD -> launchFilePicker()
            FileAction.DOWNLOAD -> if (mSelectedFile != null) {
                downloadFile(mSelectedFile!!)
            }
        }
    }

    override fun loadData() {
        mPath?.let { path ->
            val dialog: ProgressDialog = ProgressDialog(this).apply {
                setProgressStyle(ProgressDialog.STYLE_SPINNER)
                setCancelable(false)
                setMessage("Loading")
            }
            dialog.show()
            lifecycleScope.launch {
                when (val apiResult = dropboxApiWrapper.listFolders(path)) {
                    is ListFolderApiResult.Error -> {
                        dialog.dismiss()
                        Log.e(TAG, "Failed to list folder.", apiResult.e)
                        Toast.makeText(
                            this@FilesActivity,
                            "An error has occurred",
                            Toast.LENGTH_SHORT
                        )
                            .show()
                    }
                    is ListFolderApiResult.Success -> {
                        dialog.dismiss()
                        mFilesAdapter!!.setFiles(apiResult.result.entries)
                    }
                }
            }
        }
    }

    private fun downloadFile(file: FileMetadata) {
        val dialog = ProgressDialog(this).apply {
            setProgressStyle(ProgressDialog.STYLE_SPINNER)
            setCancelable(false)
            setMessage("Downloading")
        }
        dialog.show()
        lifecycleScope.launch {
            val downloadFileTaskResult = dropboxApiWrapper.download(
                this@FilesActivity.applicationContext, file
            )
            when (downloadFileTaskResult) {
                is DownloadFileTaskResult.Error -> {
                    dialog.dismiss()
                    Log.e(TAG, "Failed to download file.", downloadFileTaskResult.e)
                    Toast.makeText(
                        this@FilesActivity,
                        "An error has occurred",
                        Toast.LENGTH_SHORT
                    ).show()
                }
                is DownloadFileTaskResult.Success -> {
                    dialog.dismiss()
                    viewFileInExternalApp(
                        applicationContext,
                        downloadFileTaskResult.result
                    )
                }
            }
        }
    }

    private fun viewFileInExternalApp(context: Context, result: File) {
        val intent = Intent(Intent.ACTION_VIEW)
        val mime = MimeTypeMap.getSingleton()
        val ext = result.name.substring(result.name.indexOf(".") + 1)
        val type = mime.getMimeTypeFromExtension(ext)
        val uri: Uri = FileProvider.getUriForFile(
            context,
            context.applicationContext.packageName + ".provider", result
        )

        intent.setDataAndType(uri, type)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)

        // Check for a handler first to avoid a crash
        val manager = packageManager
        val resolveInfo = manager.queryIntentActivities(intent, 0)
        if (resolveInfo.size > 0) {
            startActivity(intent)
        }
    }

    private fun uploadFile(fileUri: String) {
        val dialog = ProgressDialog(this)
        dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER)
        dialog.setCancelable(false)
        dialog.setMessage("Uploading")
        dialog.show()
        lifecycleScope.launch {
            val localFile = UriHelpers.getFileForUri(applicationContext, Uri.parse(fileUri))

            when (val uploadedFileResult = dropboxApiWrapper.uploadFile(localFile!!, mPath!!)) {
                is FileMetadataApiResult.Error -> {
                    dialog.dismiss()
                    Log.e(TAG, "Failed to upload file.", uploadedFileResult.e)
                    Toast.makeText(
                        this@FilesActivity,
                        "An error has occurred",
                        Toast.LENGTH_SHORT
                    ).show()
                }
                is FileMetadataApiResult.Success -> {
                    val result = uploadedFileResult.fileMetadata
                    dialog.dismiss()
                    val message = result.name + " size " + result.size + " modified " +
                            DateFormat.getDateTimeInstance().format(result.clientModified)
                    Toast.makeText(this@FilesActivity, message, Toast.LENGTH_SHORT).show()

                    // Reload the folder
                    loadData()
                }
            }
        }
    }

    private fun performWithPermissions(action: FileAction) {
        if (hasPermissionsForAction(action)) {
            performAction(action)
            return
        }
        if (shouldDisplayRationaleForAction(action)) {
            AlertDialog.Builder(this)
                .setMessage("This app requires storage access to download and upload files.")
                .setPositiveButton("OK") { dialog, which -> requestPermissionsForAction(action) }
                .setNegativeButton("Cancel", null)
                .create()
                .show()
        } else {
            requestPermissionsForAction(action)
        }
    }

    private fun hasPermissionsForAction(action: FileAction): Boolean {
        for (permission in action.permissions) {
            val result = ContextCompat.checkSelfPermission(this, permission)
            if (result == PackageManager.PERMISSION_DENIED) {
                return false
            }
        }
        return true
    }

    private fun shouldDisplayRationaleForAction(action: FileAction): Boolean {
        for (permission in action.permissions) {
            if (!ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
                return true
            }
        }
        return false
    }

    private fun requestPermissionsForAction(action: FileAction) {
        ActivityCompat.requestPermissions(
            this,
            action.permissions,
            action.getCode()
        )
    }

    private enum class FileAction(vararg permissions: String) {
        DOWNLOAD(Manifest.permission.WRITE_EXTERNAL_STORAGE), UPLOAD(Manifest.permission.READ_EXTERNAL_STORAGE);

        val permissions: Array<String> = permissions.toList().toTypedArray()

        fun getCode(): Int {
            return ordinal
        }

        companion object {
            private val values = values()
            fun fromCode(code: Int): FileAction {
                require(!(code < 0 || code >= values.size)) { "Invalid FileAction code: $code" }
                return values[code]
            }
        }
    }

    companion object {
        private val TAG = FilesActivity::class.java.name
        const val EXTRA_PATH = "FilesActivity_Path"
        private const val PICKFILE_REQUEST_CODE = 1
        fun getIntent(context: Context?, path: String?): Intent {
            val filesIntent = Intent(context, FilesActivity::class.java)
            filesIntent.putExtra(EXTRA_PATH, path)
            return filesIntent
        }
    }
}